### PR TITLE
Clean up methods in "Plane" types

### PR DESCRIPTION
**Pull request by EarlinLutz**
In prior code, methods in "PlaneAltitudeEvaluator" were inconsistently supported. 

Cleanup changes:
* New abstract base class Plane3d
  *  This base class has (as abstract methods) all the methods in PlaneAltitudeEvaluator, and can have more added.
  * Multiple plane classes extend the abstract base class:
    * Plane3dByOriginAndUnitNormal
    * Plane3dByOriginAndVectors
    * Plane4d
    * ClipPlane
  * The methods in PlaneAltitudeEvaluator are implemented consistently in the derived classes.
* Implement common constructors in multiple classes
  * createOriginAndVectors
  * createOriginAndTargets
  * createFrom(Plane3d) -- Plane3dByOriginAndNormal and Plane3dByOriginAndUnitVectors will pull in any Plane3d using only calls required by the abtract base class.

### CODE REVIEW

File: core/geometry/src/geometry3d/Point3dVector3d.ts [hunk 1]
```diff
@@ -166,6 +169,14 @@ export class XYZ implements XYAndZ {
       && Geometry.isSameCoordinate(this.y, y, tol)
       && Geometry.isSameCoordinate(this.z, z, tol);
   }
+  /** Return true if this and {other + vector*scale} have equal x,y,z parts within Geometry.smallMetricDistance.
+   * * this method is useful in testing "point on ray" without explicitly constructing the projection point
+  */
```
Comment (saeeedtorabi) [inline core/geometry/src/geometry3d/Point3dVector3d.ts:174]: You have only added 2 new functions to Point3dVector3d.ts (i.e., isAlmostEqualPointPlusScaledVector and createUnitCrossProductToPoints)
Rest of the changes made to this file has (I suppose accidently) reverted the changes I made to this class. Please fix them.

File: core/geometry/src/geometry3d/Point3dVector3d.ts [hunk 2]
```diff
@@ -166,6 +169,14 @@ export class XYZ implements XYAndZ {
       && Geometry.isSameCoordinate(this.y, y, tol)
       && Geometry.isSameCoordinate(this.z, z, tol);
   }
+  /** Return true if this and {other + vector*scale} have equal x,y,z parts within Geometry.smallMetricDistance.
```
Comment (saeeedtorabi) [inline core/geometry/src/geometry3d/Point3dVector3d.ts:172]: It should be

/** 
 * Return true if this and

File: core/geometry/src/geometry3d/Point3dVector3d.ts (hunk 1 above)
Comment (EarlinLutz) [reply core/geometry/src/geometry3d/Point3dVector3d.ts:174]: I'm glad you found that clearly . ..

**Comment by dassaf4** [comment]
This PR is motivated by issue #5169.

File: core/geometry/src/geometry3d/Point3dVector3d.ts (hunk 2 above)
Comment (saeeedtorabi) [reply core/geometry/src/geometry3d/Point3dVector3d.ts:169]: this is still not addressed

**Review by dassaf4 — APPROVED** [review]

File: core/geometry/src/clipping/ClipPlane.ts [hunk 3]
```diff
@@ -572,4 +572,8 @@ export class ClipPlane implements Clipper, PlaneAltitudeEvaluator, PolygonClippe
     ClipUtilities.captureOrDrop(newInside, 3, insideFragments, arrayCache);
     ClipUtilities.captureOrDrop(newOutside, 3, outsideFragments, arrayCache);
   }
+  public projectPointToPlane(spacePoint: Point3d, result?: Point3d): Point3d {
```
[the remark this answers is not in the document]
Comment (EarlinLutz) [reply core/geometry/src/clipping/ClipPlane.ts:591]: done.

File: core/geometry/src/geometry3d/Ray3d.ts [hunk 4]
```diff
@@ -54,10 +54,23 @@ export class Ray3d implements BeJSONFunctions {
...
   public isAlmostEqual(other: Ray3d): boolean {
     return this.origin.isAlmostEqual(other.origin) && this.direction.isAlmostEqual(other.direction);
   }
+  /** Test for nearly equal rays, allowing origin float and direction scaling.
+   * * Use [[isAlmostEqual]] to require member-by-member comparison.
+  */
+  public isAlmostEqualPointSet(other: Ray3d): boolean {
+    if (!this.direction.isParallelTo(other.direction, true))
+      return false;
+    // In exact math, it is not possible for one origin to be on the other ray but not vice versa.  But we'll test both ways.
```
[the remark this answers is not in the document]
Comment (EarlinLutz) [reply core/geometry/src/geometry3d/Ray3d.ts:70]: this.origin on other added.

File: core/geometry/src/geometry3d/Plane3dByOriginAndUnitNormal.ts [hunk 5]
```diff
@@ -254,6 +280,11 @@ export class Plane3dByOriginAndUnitNormal implements BeJSONFunctions, PlaneAltit
   public projectPointToPlane(spacePoint: Point3d, result?: Point3d): Point3d {
     return spacePoint.plusScaled(this._normal, -this._normal.dotProductStartEnd(this._origin, spacePoint), result);
   }
-  /** Returns true of spacePoint is within distance tolerance of the plane. */
-  public isPointInPlane(spacePoint: Point3d): boolean { return Geometry.isSmallMetricDistance(this.altitude(spacePoint)); }
+  /** Returns true of spacePoint is within distance tolerance of the plane.
+   * @remark This logic is identical to the abstract Plane3d but avoids a level of function call.
```
[the remark this answers is not in the document]
Comment (pmconne) [reply core/geometry/src/geometry3d/Plane3dByOriginAndUnitNormal.ts:284]: `@remark` is not a valid tag.

**Review by dassaf4 — APPROVED** [review]

File: common/api/core-geometry.api.md [hunk 6]
```diff
@@ -4150,6 +4159,7 @@ export interface PlaneAltitudeEvaluator {
     normalX(): number;
     normalY(): number;
     normalZ(): number;
+    projectPointToPlane(spacePoint: Point3d, result?: Point3d): Point3d;
```
[the remark this answers is not in the document]
Comment (EarlinLutz) [reply common/api/core-geometry.api.md:4162]: I'll move that down to the base class.

File: core/geometry/src/geometry4d/Point4d.ts [hunk 7]
```diff
@@ -343,6 +352,18 @@ export class Point4d implements BeJSONFunctions {
     }
     return undefined;
   }
+  /**
+   * * Return xyz projection of spacePoint to the plane of the DPoint4d (understood as coefficients, not point coordinates)
+   * * If the xyz part of this are ll zero, (a clone of) spacePOint is returned.
+   */
+  public projectPointToPlane(spacePoint: Point3d, result?: Point3d): Point3d {
+    const h = this.altitude(spacePoint);
+    const nn = this.magnitudeSquaredXYZ();
```
[the remark this answers is not in the document]
Comment (EarlinLutz) [reply core/geometry/src/geometry4d/Point4d.ts:363]: The toPlane3d bit has been there -- not a new bug (if it is a bug -- there may be a subsequent normalization that undoes it.) 
BUT -- the body is changed to a one-liner call to Plane3dByOriginandNormal.createFrom().

File: common/api/core-geometry.api.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (EarlinLutz) [reply common/api/core-geometry.api.md]: having projectPointToPlane in the interface was indeed incorrect and unnecessary.   Corrected now.

File: core/geometry/src/test/interfaceTest.ts.examplesOfInterfaceAndAbstractClassEvolution [hunk 8]
```diff
@@ -0,0 +1,90 @@
+/*---------------------------------------------------------------------------------------------
+* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+* See LICENSE.md in the project root for license terms and full copyright notice.
+*--------------------------------------------------------------------------------------------*/
+
+// import { expect } from "chai";
+
+// Experiments with adding a method to a class (a) as part of a larger interface and (b) as part of a new abstract base class.
```
[the remark this answers is not in the document]
Comment (EarlinLutz) [reply core/geometry/src/test/interfaceTest.ts.examplesOfInterfaceAndAbstractClassEvolution]: done

**Review by dassaf4 — APPROVED** [review]

File: core/geometry/src/geometry4d/Point4d.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (dassaf4) [reply core/geometry/src/geometry4d/Point4d.ts:363]: Nice

**Review by DanRod1999 — APPROVED** [review]

**Pull request merged**